src/dftbp/type/integral.F90

```diff
@@ -0,0 +1,61 @@
+!--------------------------------------------------------------------------------------------------!
+!  DFTB+: general package for performing fast atomistic simulations                                !
+!  Copyright (C) 2006 - 2021  DFTB+ developers group                                               !
+!                                                                                                  !
+!  See the LICENSE file for terms of usage and distribution.                                       !
+!--------------------------------------------------------------------------------------------------!
+
+!> Data types to handle overlap related integrals
+module dftbp_type_integral
+  use dftbp_common_accuracy, only : dp
+  implicit none
+
+  private
+  public :: TIntegral, TIntegral_init
+
+
+  !> Container to store overlap related integrals
+  type :: TIntegral
```

Comment:
For multi-determinantal cases (Delta SCF, REKS, CI, ...), the overlap would be the same for different instances of the hamiltonian. Presumably OK if they are processed sequentially in serial, as uses only one copy of this structure (and parallel doesn't care too much). Suggests either a) live with the duplication b) pointers :( or c) put the overlap in its own structure.
Also, is the reason not to include H0 in the structure to avoid making this model specific?

Reply:
Delta SCF and REKS are bypassing the issue with multiple Hamiltonians either by doing serial processing (Delta SCF) or by allocating their own sparse arrays (REKS). Splitting the Hamiltonian to a separate data structure, maybe together with the density matrix would help to create a more general CI framework in DFTB+. Still, I don't see a way to make such change immediately useful for Delta SCF or REKS.

The other issue is that the core Hamiltonian can currently not be included in this data structure since the real time propagation is using a local copy and I haven't looked through the whole code to check whether this is actually necessary.

Reply:
The RT propagation is updating H0 when it changes geometry, but for convenience just re-used the initial ground state H0 when initially implemented. As the code flow won't go back to a ground state calc after running propagation (?) the ground state H0 can be removed and only one structure used.

Reply:
Sounds good to me, I can check if I'm able to reduce it a single instance without breaking anything. Maybe I can give it a stab later today, but first multipoles.

Reply:
Yes, both Delta SCF and REKS are sequential in working with determinants. They should get refactored to run them in parallel for the MPI build, but probably kept sequential for the openMP/serial case (which is where saving the memory for multiple overlap copies would matter).